orttraining/orttraining/training_ops/cuda/loss/softmax_cross_entropy_loss_impl.cc

```diff
@@ -157,13 +165,26 @@ Status SoftmaxCrossEntropyLoss<T, Tin>::ComputeInternal(OpKernelContext* ctx) co
 
   if (reduction_ != ReductionType::NONE) {
     // ReduceSum on loss_per_sample
-    ORT_RETURN_IF_ERROR(reduce_sum(
-        Stream(),
-        tmp_loss_sample_buffer,
-        total_loss_data,
-        static_cast<int>(N_D),
-        reduction_buffer.get(),
-        buffer_size));
+    if (!std::is_same<T, TBuf>::value) {
+      ORT_RETURN_IF_ERROR(reduce_sum(
+          Stream(),
+          reinterpret_cast<CudaT*>(tmp_loss_sample_buffer),
+          normalize_factor_data.get(),
+          static_cast<int>(N_D),
+          reduction_buffer.get(),
+          buffer_size));
+      // Convert from accumulation type to output type
+      // Need a better way to do this cast as this is a single element cast
+      Impl_Cast<TBuf, CudaT>(Stream(), normalize_factor_data.get(), reinterpret_cast<CudaT*>(total_loss_data), 1);
```

Comment:
How about extending reduce_sum to support this case rather than using Impl_Cast here?